### DISCUSSION TITLE
Questions about development.

### DISCUSSION THREAD

**Discussion by andrewfer000** [SoftEtherVPN/SoftEtherVPN#1431] [General]
Hello SE devs, 

I have been studying your code and learning C for awhile now and I would like some guides on contributing such as

What are your setups for programmimg SoftEther? What OSes and IDEs do you use? What's the best way to develop on Linux?

I know this is a stupid question, but what does the `->` mean in the code? I see it alot but I never used it when I practiced C programming.

What C programming concepts are the most important for developing SE? (i.e. Classes, Pointers, Structures, I/O, etc.)

Where should I start to get IPv6 working on the IPC stack protocols? (OpenVPN/WireGuard)

Would anyone want to mentor me?

Thanks Devs, let me know.

**Comment by davidebeatrici** [comment]
Hello, thanks for your interest!

> What are your setups for programmimg SoftEther? What OSes and IDEs do you use? What's the best way to develop on Linux?

I personally use KDevelop on Linux. When I have to develop something Windows-specific I use Visual Studio.

> I know this is a stupid question, but what does the `->` mean in the code? I see it alot but I never used it when I practiced C programming.

It's used to dereference a pointer. Example:

```c
#include <stdio.h>
#include <stdlib.h>

typedef struct Struct
{
	int var;
};

int main()
{
	Struct *pointer = malloc(sizeof(Struct));

	pointer->var = 1;
	printf("var: %d\n", pointer->var);

	free(pointer);

	return 0;
}

```

> What C programming concepts are the most important for developing SE? (i.e. Classes, Pointers, Structures, I/O, etc.)

Classes only exist in C++. Good understanding of pointers and structures is mandatory for basically the entire codebase.

As for I/O: it depends on the kind and the specific part of code you're working on.

> Where should I start to get IPv6 working on the IPC stack protocols? (OpenVPN/WireGuard)

IPC in theory already supports IPv6, meaning that it's probably just a matter of enabling it and verify that it works correctly.

See https://github.com/SoftEtherVPN/SoftEtherVPN/blob/master/src/Cedar/IPC.c.

> Would anyone want to mentor me?

Yes, feel free to ask anything in this discussion.

**Comment by andrewfer000** [comment]
@davidebeatrici 
Thanks for your reply, I'll start reading up on advanced C concepts such as the pointers and structures. I know the basics pretty well but I don't fully understand everything I need to make a valuable contribution yet especially to a program that has such a strict rules on memory management. Once I get my main laptop back from the repair shop I'll try to get SoftEther in Kdevelop. If you have any specific advice or guides for getting kdevelop up and running for SE development let me know.

However, I don't know how to use to git very well so if I do make a valueble contribution I'll just upload it to my GitHub account and have you or another developer do the PR stuff. 

In any case we'll see how things go when I get my laptop back. I look forward to helping out this project and gaining new experience.


Btw, if IPv6 support is already on the IPC stack and it just needs to be enabled for the protocols, I wouldn't mind doing the testing if you (or another dev) have the time and knowledge to get it done quicker and cleaner than I would (if I even can do it). This is one of the major features I believe needs to be implemented to compete better with the other VPN server products out there. 

Also, by any chance do you have that API documentation that's an HTML document hosted somewhere? It's kind of hard to read and understand it in html code.

**Reply by davidebeatrici** [reply]
> If you have any specific advice or guides for getting kdevelop up and running for SE development let me know.

You only have to open the CMake project in it to get started.

For quick debugging you can add a launch target, such as `vpntest s` for the server.

> However, I don't know how to use to git very well so if I do make a valueble contribution I'll just upload it to my GitHub account and have you or another developer do the PR stuff.

Sure, no problem.

> Btw, if IPv6 support is already on the IPC stack and it just needs to be enabled for the protocols, I wouldn't mind doing the testing if you (or another dev) have the time and knowledge to get it done quicker and cleaner than I would (if I even can do it). This is one of the major features I believe needs to be implemented to compete better with the other VPN server products out there.

I will work on implementing IPv6 support for WireGuard as soon as possible.

> Also, by any chance do you have that API documentation that's an HTML document hosted somewhere? It's kind of hard to read and understand it in html code.

It's actually accessible when you run the server. Also, you can open the file in a browser after cloning the repository.

**Comment by andrewfer000** [comment]
@davidebeatrici So I'm going to try to get this up and running on my temp laptop just to see what it's like. Are there any small bugs that I can tackle kind of like a starting exercise? Are those Coverity things I've been seeing all over the issue tracker need fixing?

Also do I need to run `git submodule init && git submodule update` or  `./configure` before opening it in Kdevelop? Or do I do those after I'm done making changes and ready to build?

I'm sorry that I have so many questions. Most of my experience is in small C++ and C projects where I used the command line for everything. So working with such a large project is a new experience for me. Also how do I set up launch targets? Are they preconfigured or do I need to set them up in some kind of config file? 

Once again thanks for putting up with me :).

So I tried to import to Kdevelop with some success. But I get this error message at top
```
Failed to configure project 'SoftEtherVPN' (error message: Configuration failed.). As a result, KDevelop's code understanding will likely be broken.

To fix this issue, please ensure that the project's CMakeLists.txt files are correct, and KDevelop is configured to use the correct CMake version and settings. Then right-click the project item in the projects tool view and click 'Reload'.
```

**Reply by davidebeatrici** [reply]
> Are there any small bugs that I can tackle kind of like a starting exercise? Are those Coverity things I've been seeing all over the issue tracker need fixing?

Yes.

> Also do I need to run `git submodule init && git submodule update` or  `./configure` before opening it in Kdevelop? Or do I do those after I'm done making changes and ready to build?

Execute `git submodule update --init --recursive` and then open the project in KDevelop.

> I'm sorry that I have so many questions. Most of my experience is in small C++ and C projects where I used the command line for everything. So working with such a large project is a new experience for me. Also how do I set up launch targets? Are they preconfigured or do I need to set them up in some kind of config file?

You can use the command line to build SoftEther VPN as well, if you like. An IDE is just a commodity.

![](https://user-images.githubusercontent.com/5897523/125571290-f44a18c8-0fa6-4fcc-ae27-b394fa93090e.png)
![](https://user-images.githubusercontent.com/5897523/125571450-1a572975-50f5-4aa2-9608-8cc6d0687f58.png)

> Once again thanks for putting up with me :).

No worries!

> So I tried to import to Kdevelop with some success. But I get this error message at top
> ```
> Failed to configure project 'SoftEtherVPN' (error message: Configuration failed.). As a result, KDevelop's > code understanding will likely be broken.
>
> To fix this issue, please ensure that the project's CMakeLists.txt files are correct, and KDevelop is configured to use the correct CMake version and settings. Then right-click the project item in the projects tool view and click 'Reload'.
> ```

Check CMake's output.

**Comment by andrewfer000** [comment]
Thanks for the infos. I'm going to go to bed now, but I'll mess around with the IDE stuff a little more tomorrow after I'm done my school work and hopefully I can try to do something.  

By the way, about the Coverity things, how do I know if I fixed them or not?

**Reply by davidebeatrici** [reply]
You have to wait for the next Coverity scan.

**Reply by chipitsine** [reply]
@andrewfer000 , Coverity hides issue definitions. I added several last issues to github issue. If you wish, you can review them. Also, if that helps, I will add all existing coverity issues ( around 50) to github issues

**Reply by andrewfer000** [reply]
Let me take a look at some of the ones allready here first. I'm still feeling my way around the project. It's very large especially for someone with not much expreince. I should get my main laptop back by tomorrow at the latest so then I can really get to work.

I'm very inexperienced with large projects so if you give me some tips on where to start and how to navigate this project it'll be great.

**Comment by andrewfer000** [comment]
Just got my laptop back. however when I try to build SoftEther on it I get a dependency error for libsodium-dev. Apt says  libsodium23 is a dependency for it but I have installed. If I try to remove libsodium23 it's going to break a lot of my programs. So I'm going to have to put programming on pause for a little bit until I can get something like a lxc container set up for development. Which I can absolutely do just as long as SoftEther runs fine in a container environment.

**Reply by chipitsine** [reply]
What is your Ubuntu version?

**Reply by andrewfer000** [reply]
> What is your Ubuntu version?

20.04 running Kernel 5.10.

**Reply by davidebeatrici** [reply]
Removing `libsodium23` through APT will in turn remove all packages that depend on it.

`libsodium-dev` contains the development files required to build programs that make use of the library.

What's the output when you try to install the package?

**Reply by andrewfer000** [reply]
```
The following packages have unmet dependencies:
 libsodium-dev : Depends: libsodium23 (= 1.0.18-1) but 1.0.18-1+ubuntu19.10.1+deb.sury.org+1 is to be installed
E: Unable to correct problems, you have held broken packages.
```

But the thing is I already have libsodium23  installed.

**Reply by davidebeatrici** [reply]
It's trying to install the package for Ubuntu 19.10, check `/etc/apt/sources.list` and `/etc/apt/sources.list.d`.

**Reply by andrewfer000** [reply]
I moved and remade `/etc/apt/sources.list.d/` and ran `sudo apt update`
```
owner@owner-Zephyrus-G:~$ sudo apt update
Hit:1 http://us.archive.ubuntu.com/ubuntu focal InRelease
Hit:2 http://security.ubuntu.com/ubuntu focal-security InRelease
Hit:3 http://us.archive.ubuntu.com/ubuntu focal-updates InRelease
Hit:4 http://us.archive.ubuntu.com/ubuntu focal-backports InRelease
Reading package lists... Done
Building dependency tree       
Reading state information... Done
All packages are up to date.
owner@owner-Zephyrus-G:~$ sudo apt install libsodium-dev
Reading package lists... Done
Building dependency tree       
Reading state information... Done
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:

The following packages have unmet dependencies:
 libsodium-dev : Depends: libsodium23 (= 1.0.18-1) but 1.0.18-1+ubuntu19.10.1+deb.sury.org+1 is to be installed
E: Unable to correct problems, you have held broken packages.
```

Still got the same error.

**Reply by andrewfer000** [reply]
Part of me want's to give up... I want to program but I hate Linux dependency shit. I love Linux but this is a huge load of BS!

**Reply by davidebeatrici** [reply]
You have the wrong repositories in `/etc/apt/sources.list`, check the file's content.

**Reply by andrewfer000** [reply]
I'm working on getting my LXC container setup. I am able to run full graphical applications in it as if it was running like a normal program on my OS. It works great on Chrome so I don't see why Git and Kdevelop won't work. I'll try and get that all set up tomorrow.

**Reply by andrewfer000** [reply]
I'm going to compare the file with the one on my VPN Server. Since I can compile SoftEther on that it should be able to get all the dependencies I need.

**Reply by davidebeatrici** [reply]
You can't keep your Ubuntu setup like that, it's possible it will eventually break because of the wrong repository.

Could you send the file's content here, please?

**Reply by andrewfer000** [reply]
My `/etc/apt/sources.list` now matches the one on my server but still no luck. Hopefully the LXC container doesn't give me issues or else I'm going to have to use a sluggish VM.

```
# See http://help.ubuntu.com/community/UpgradeNotes for how to upgrade to
# newer versions of the distribution.
deb http://us.archive.ubuntu.com/ubuntu focal main restricted
# deb-src http://us.archive.ubuntu.com/ubuntu focal main restricted

## Major bug fix updates produced after the final release of the
## distribution.
deb http://us.archive.ubuntu.com/ubuntu focal-updates main restricted
# deb-src http://us.archive.ubuntu.com/ubuntu focal-updates main restricted

## N.B. software from this repository is ENTIRELY UNSUPPORTED by the Ubuntu
## team. Also, please note that software in universe WILL NOT receive any
## review or updates from the Ubuntu security team.
deb http://us.archive.ubuntu.com/ubuntu focal universe
# deb-src http://us.archive.ubuntu.com/ubuntu focal universe
deb http://us.archive.ubuntu.com/ubuntu focal-updates universe
# deb-src http://us.archive.ubuntu.com/ubuntu focal-updates universe

## N.B. software from this repository is ENTIRELY UNSUPPORTED by the Ubuntu
## team, and may not be under a free licence. Please satisfy yourself as to
## your rights to use the software. Also, please note that software in
## multiverse WILL NOT receive any review or updates from the Ubuntu
## security team.
deb http://us.archive.ubuntu.com/ubuntu focal multiverse
# deb-src http://us.archive.ubuntu.com/ubuntu focal multiverse
deb http://us.archive.ubuntu.com/ubuntu focal-updates multiverse
# deb-src http://us.archive.ubuntu.com/ubuntu focal-updates multiverse

## N.B. software from this repository may not have been tested as
## extensively as that contained in the main release, although it includes
## newer versions of some applications which may provide useful features.
## Also, please note that software in backports WILL NOT receive any review
## or updates from the Ubuntu security team.
deb http://us.archive.ubuntu.com/ubuntu focal-backports main restricted universe multiverse
# deb-src http://us.archive.ubuntu.com/ubuntu focal-backports main restricted universe multiverse

## Uncomment the following two lines to add software from Canonical's
## 'partner' repository.
## This software is not part of Ubuntu, but is offered by Canonical and the
## respective vendors as a service to Ubuntu users.
# deb http://archive.canonical.com/ubuntu focal partner
# deb-src http://archive.canonical.com/ubuntu focal partner

deb http://us.archive.ubuntu.com/ubuntu focal-security main restricted
# deb-src http://us.archive.ubuntu.com/ubuntu focal-security main restricted
deb http://us.archive.ubuntu.com/ubuntu focal-security universe
# deb-src http://us.archive.ubuntu.com/ubuntu focal-security universe
deb http://us.archive.ubuntu.com/ubuntu focal-security multiverse
#deb-src http://us.archive.ubuntu.com/ubuntu focal-security multiverse

```

**Reply by davidebeatrici** [reply]
Have you run `sudo apt update` after updating the file?

**Reply by andrewfer000** [reply]
Yes. Btw I just fixed it by running `sudo apt install --reinstall libsodium23=1.0.18-1` and `sudo apt install libsodium-dev`. Now it's installed.

**Reply by andrewfer000** [reply]
Everything looks good in Kdevelop now. No more errors. 

Let's just hope I didn't break any of my other programs.

**Comment by andrewfer000** [comment]
Okay now that my Dependency Hell has been resolved and I am now able to build SE inside of Kdevelop. Let's discuss some real programming stuff now. 

This project is huge... Where should I start?
